go/src/koding/klient/machine/mount/index/index.go

```diff
@@ -0,0 +1,328 @@
+package index
+
+import (
+	"crypto/sha1"
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+// Entry represents a single file registered to index.
+type Entry struct {
+	Name  string      `json:"name"`  // The relative name of the file.
+	CTime int64       `json:"ctime"` // Metadata change time since EPOCH.
+	MTime int64       `json:"mtime"` // File data change time since EPOCH.
+	Mode  os.FileMode `json:"mode"`  // File mode and permission bits.
+	Size  int64       `json:"size"`  // Size of the file.
+	SHA1  []byte      `json:"sha1"`  // SHA-1 hash of file content.
+}
+
+// NewEntryFile creates new Entry from a file stored under path argument.
+// Info is optional and, if given, should store LStat result on the given file.
+func NewEntryFile(root, path string, info os.FileInfo) (e *Entry, err error) {
+	if info == nil {
+		if info, err = os.Lstat(path); err != nil {
+			return nil, err
+		}
+	}
+
+	// Get relative file name.
+	name, err := filepath.Rel(root, path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compute file's SHA-1 sum.
+	var sum []byte
+	if !info.IsDir() {
+		if sum, err = readSHA1(path); err != nil {
```

Comment:
I'd like to keep SHA-1 here because I have some interest in keeping hashes consistent with these stored in git index(which means that I need SHA-1).

I have (long-term) plan to add another `Syncer` interface implementation. Lets call it `GitSync` that can detect git repository in mounted folder and improve git operations on it. Eg.

User runs `git checkout my_other_branch` on mounted directory. This can create a lot of FUSE changes and require as much of Rsync syncer work. Now, assuming we have `GitSync` that can examine git index/reflog/etc. and detect eg. checkout. Thus, it will be able to just run `git checkout mu_other_branch` on remote machine without copying anything.

These are just plans but, if we ever start implementing this, it would be nice to have SHA-1 sums in order to perform sanity checks on git index.